### DISCUSSION TITLE
Move Settings Across Environments

### DISCUSSION THREAD

**Discussion by oluatte** [logto-io/logto#4821] [Q&A]
Say you've got LogTo setup the way you like it in your dev environment. You've got the right applications, apis, roles, permissions, branding etc. Now you want to move to a staging or production environment. 

How are people moving the settings across?

- SQL scripts that are run in each environment?
- Any other ways to do this?

**Comment by simeng-li** [comment]
Hi @mayoatte thanks for the suggestion. It's indeed a valuable one. We are doing some related research as well. 

To better assist you and help us understand the customer needs could you provide some more details:

- Are you utilizing the open-source version or our Logto cloud service?
- Please share more specific details about the resources and data you wish to transport from a development tenant to a production tenant. 
- Are you expecting all "cloned" resources to retain the same ID (references at the client side), or are you interested in cloning only configuration settings? 
- Additionally, let us know if you prefer to "clone" everything or per different data resource types.

Thanks.

**Reply by oluatte** [reply]
Howdy. Thanks for the quick response.

- I'm using the open-source version. I'm trying out a few identity providers to see which one has the right feature set for our project (B2B2C).
- It would be valuable to move APIs, Applications, Roles, Permissions, Sign In Experience settings, maybe even connector settings.
- I don't expect the cloned resources to maintain the same id.  
- Cloning everything is fine.

Another (better) approach entirely would be allowing users to supply the configuration for each environment via a file or some other means. This would allow us to configure the IDP in the same infra as code style as we do other things.

Thanks for listening.

**Comment by Petteroe** [comment]
I am in a situation where I have played around with the OSS version in a dev environment for a while and have started to structure a relatively complex setup with organizations, applications, roles and permissions. It would be a time-saver to be able to promote this to a new environment as an export->import thing. Is there any more support for this now than there was in Nov 2023?

**Comment by bitpickle** [comment]
The best way I found to do this was to create a script using the Logto API. I'm using this approach just to set up the environment (roles, applications, etc.) but you can go further and create the users, since Logto accepts your existing hashes in the user creation endpoint.

In short, it is possible, but it would be nice to have an official way to do this.

Another possibility, thinking about the open source version, is to simply export your database, but I have never tested it.

P.S.: I'm using [OpenAPI Generator](https://openapi-generator.tech/) to generate a SDK.
